dhall/src/Dhall/Import/Types.hs

```diff
@@ -33,28 +34,35 @@ import qualified Dhall.Context
 import qualified Data.Map      as Map
 import qualified Data.Text
 
+-- | A fully 'chained' import, i.e. if it contains a relative path that path is
+--   relative to the current directory. If it is a remote import with headers
+--   those are well-typed and in normal form. Note that the type itself isn't
+--   precise enough to capture either of these invariants.
```

Comment:
Even though the type might not capture those invariants, you could make it so that the API exposed by the module captures those invariants.  The module could expose only safe primitives for building and transforming `Chained` imports

Reply:
That was the original idea. I messed that up when trying to get dhall-lsp-server to compile.